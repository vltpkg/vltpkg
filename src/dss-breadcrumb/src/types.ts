/**
 * The specificity of a breadcrumb, used for sorting.
 */
export type BreadcrumbSpecificity = {
  idCounter: number
  commonCounter: number
}

/**
 * Options for the higher-level comparing breadcrumbs function.
 */
export type InternalModifierComparatorOptions = {
  range?: string
}

/**
 * Options for comparing breadcrumbs.
 */
export type ModifierComparatorOptions = {
  semver?: string
}

/**
 * A valid item of a given breadcrumb.
 */
export type ModifierBreadcrumbItem = {
  comparator: (opts: ModifierComparatorOptions) => boolean
  name?: string
  value: string
  type: string
  importer: boolean
  prev: ModifierBreadcrumbItem | undefined
  next: ModifierBreadcrumbItem | undefined
}

/**
 * A breadcrumb is a linked list of items, where
 * each item has a value and a type.
 */
export interface ModifierBreadcrumb
  extends Iterable<ModifierBreadcrumbItem> {
  clear(): void
  comment: string | undefined
  first: ModifierBreadcrumbItem
  last: ModifierBreadcrumbItem
  single: boolean
  specificity: BreadcrumbSpecificity
  interactive: () => ModifierInteractiveBreadcrumb
}

/**
 * An interactive breadcrumb that holds state on what is the current item.
 */
export type ModifierInteractiveBreadcrumb = {
  current: ModifierBreadcrumbItem | undefined
  done: boolean
  next: () => ModifierInteractiveBreadcrumb
}
