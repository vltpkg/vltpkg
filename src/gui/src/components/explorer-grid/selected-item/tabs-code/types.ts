import type { FsItem } from '@/lib/fetch-fs'

export type FsItemWithNone = Omit<
  Pick<FsItem, 'type' | 'name' | 'size'>,
  'type'
> & {
  type: FsItem['type'] | 'none'
}

export interface ItemAlerts {
  title: string
  description: string
}

export interface Error {
  origin: string
  cause: string
}

export type View = 'code' | 'preview'

export type Language =
  | 'markdown'
  | 'typescript'
  | 'javascript'
  | 'jsx'
  | 'tsx'
  | 'json'
  | 'css'
  | 'html'
  | 'text'

export type SortingDirection = 'asc' | 'desc' | 'unsorted'

export type FsItemSortKey = 'name' | 'type' | 'size' | null
