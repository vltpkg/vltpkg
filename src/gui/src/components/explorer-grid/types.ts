import type { QueryResponseEdge } from '@vltpkg/query'

/**
 * A looser representation of an edge, it contains optional properties that
 * are meant to be used alongside importer nodes where they might be missing.
 */
export type EdgeLoose = Pick<QueryResponseEdge, 'name'> & {
  type?: QueryResponseEdge['type']
  spec?: QueryResponseEdge['spec']
  from?: QueryResponseEdge['from']
  to?: QueryResponseEdge['to']
}

/**
 * Data to be presented in the UI in the form of Cards. This is used for
 * both the main result column, selected node and side panel views.
 */
export type GridItemData = EdgeLoose & {
  /**
   * The original dependency name as defined in its parent manifest.
   */
  depName?: string
  /**
   * An index value shared between installed and uninstalled dependencies
   * to keep track of the order they were added to the UI.
   */
  depIndex?: number
  /**
   * A unique identifier for the item. This will not translate to `node.to.id`
   * but rather concatenate names / ids to create a unique id per item.
   * Used as the unique `key` for React components.
   */
  id: string
  /**
   * Whether the current result interface is showing only results for a single
   * node. Used for dynamically building queries when clicking the UI.
   */
  sameItems?: boolean
  /**
   * How many resulting edges this item represents.
   */
  size: number
  /**
   * Whether this item is representing multiple edge results.
   */
  stacked: boolean
  /**
   * A title to be displayed in the Card UI.
   */
  title: string
  /**
   * A version to be used in the UI.
   */
  version: string
  /**
   * A list of labels to be displayed in the Card UI.
   */
  labels?: string[]
}

/**
 * Options used to create new Card grid items to be displayed in the UI.
 */
export type GridItemOptions = {
  item: GridItemData
  dependencies?: boolean
  highlight?: boolean
  selected?: boolean
  side?: boolean
}
