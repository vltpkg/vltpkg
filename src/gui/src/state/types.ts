import {
  type LockfileData,
  type GraphLike,
  type NodeLike,
  type EdgeLike,
} from '@vltpkg/graph'
import { type Query } from '@vltpkg/query'
import { type SpecOptionsFilled } from '@vltpkg/spec/browser'
import { type Integrity, type Manifest } from '@vltpkg/types'

export type Action = {
  updateActiveRoute: (route: State['activeRoute']) => void
  updatePreviousRoute: (route: State['activeRoute']) => void
  updateDashboard: (dashboard: State['dashboard']) => void
  updateGraph: (graph: State['graph']) => void
  updateQ: (q: State['q']) => void
  updateQuery: (query: State['query']) => void
  updateEdges: (edges: State['edges']) => void
  updateErrorCause: (errorCause: State['errorCause']) => void
  updateHasDashboard: (hasDashboard: State['hasDashboard']) => void
  updateLinePositionReference: (position: number) => void
  updateNodes: (nodes: State['nodes']) => void
  updateProjectInfo: (projectInfo: State['projectInfo']) => void
  updateSelectedNode: (node: State['selectedNode']) => void
  updateSpecOptions: (specOptions: State['specOptions']) => void
  updateStamp: () => void
  updateTheme: (theme: State['theme']) => void
  reset: () => void
  saveQuery: (item: SavedQuery) => void
  updateSavedQuery: (savedQuery: SavedQuery) => void
  deleteSavedQueries: (savedQueries: SavedQuery[]) => void
  saveQueryLabel: (queryLabel: QueryLabel) => void
  updateSavedQueryLabel: (queryLabel: QueryLabel) => void
  deleteSavedQueryLabels: (queryLabels: QueryLabel[]) => void
}

/**
 * Project information present in the transfer data object.
 */
export type ProjectInfo = {
  /**
   * Other tools used by this project, such as runtime and package mangager.
   */
  tools: DashboardTools[]
  /**
   * `true` if this package node_modules folder was installed
   * using the vlt client.
   */
  vltInstalled?: boolean
}

/**
 * Transfer data object used to send data from the cli.
 */
export type TransferData = {
  importers: RawNode[]
  lockfile: LockfileData
  projectInfo: ProjectInfo
}

export type RawNode = {
  importer: boolean
  id: string
  name: string
  version: string
  location: string
  manifest: Manifest
  projectRoot?: string
  integrity?: Integrity
  resolved?: string
  dev: boolean
  optional: boolean
}

/**
 * The main state object for the graph explorer.
 */
export type State = {
  /**
   * The current location.pathname (e.g. route) in the app.
   */
  activeRoute: string
  /**
   * The last route in the app.
   */
  previousRoute: string
  /**
   * List of projects to be displayed in the dashboard.
   */
  dashboard?: DashboardData
  /**
   * Current graph to be explored.
   */
  graph?: GraphLike
  /**
   * List of selected edges returned after querying the graph.
   */
  edges: EdgeLike[]
  /**
   * An informative message to be displayed when an error occurs.
   */
  errorCause: string
  /**
   * Whether the dashboard is enabled or not.
   */
  hasDashboard: boolean
  /**
   * A reference value to properly draw connections between nodes in the graph.
   */
  linePositionReference: number
  /**
   * List of selected nodes returned after querying the graph.
   */
  nodes: NodeLike[]
  /**
   * Information about the current project being explored.
   */
  projectInfo: ProjectInfo
  /**
   * Holds the reference to the {@link Query} instance object.
   */
  q?: Query
  /**
   * The query string typed by the user in the interface.
   */
  query: string
  // TODO: remove selectedNode as it's unused
  /**
   * Reference to a currently selected node.
   */
  selectedNode?: NodeLike
  /**
   * Spec options used for the current graph.
   */
  specOptions?: SpecOptionsFilled
  /**
   * A random string to control when graph data should be reloaded.
   */
  stamp: string
  /**
   * Store the current theme value.
   */
  theme: 'light' | 'dark'
  /**
   * Saved queries in localStorage.
   */
  savedQueries?: SavedQuery[]
  /**
   * Saved labels used for query tags in localStorage.
   */
  savedQueryLabels?: QueryLabel[]
}

export type DashboardTools =
  | 'vlt'
  | 'node'
  | 'deno'
  | 'bun'
  | 'npm'
  | 'pnpm'
  | 'yarn'
  | 'js'

export type DashboardLocation = {
  /**
   * The file system path to the directory.
   */
  path: string
  /**
   * A shortened, human readable path to be displayed in the dashboard.
   */
  readablePath: string
}

export type DashboardData = {
  /**
   * The reference current working directory.
   */
  cwd: string
  /**
   * The app version.
   */
  buildVersion: string
  /**
   * The default author name to be used when creating new projects.
   */
  defaultAuthor: string
  /**
   * A list of projects to be displayed in the dashboard.
   */
  projects: DashboardDataProject[]
  /**
   * A list of directories that are read to find projects
   * to display in the dashboard. New projects can also be created
   * in these folders.
   */
  dashboardProjectLocations: DashboardLocation[]
}

export type QueryLabel = {
  id: string
  color: Color
  name: string
  description: string
}

export type SavedQuery = {
  id: string
  name: string
  context: string
  query: string
  dateCreated: string
  dateModified: string
  labels?: QueryLabel[]
}

export type DashboardDataProject = {
  name: string
  readablePath: string
  path: string
  manifest: Manifest
  tools: DashboardTools[]
  mtime?: number | null
}

export type Color = `#${string}`
