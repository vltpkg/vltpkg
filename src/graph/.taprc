extends: ../../tap-config.yaml
disable-coverage: true