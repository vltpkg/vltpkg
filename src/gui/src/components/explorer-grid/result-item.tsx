import { type MouseEvent } from 'react'
import { stringifyNode } from '@vltpkg/graph/browser'
import { useGraphStore } from '@/state/index.js'
import { Badge } from '@/components/ui/badge.jsx'
import { Card, CardHeader, CardTitle } from '@/components/ui/card.jsx'
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from '@/components/ui/tooltip.jsx'
import { labelClassNamesMap } from './label-helper.js'
import { type GridItemData, type GridItemOptions } from './types.js'

export type ResultItemClickOptions = {
  item: GridItemData
  query: string
  updateQuery: (query: string) => void
}

const onResultItemClick =
  ({ item, query, updateQuery }: ResultItemClickOptions) =>
  (e: MouseEvent) => {
    e.preventDefault()
    if (!item.to) return
    let newQuery = ''
    if (item.stacked) {
      const name = item.to.name ? `[name="${item.to.name}"]` : ''
      const version =
        item.to.version ? `[version="${item.to.version}"]` : ''
      newQuery = `${query.trim()}${name}${version}`
    } else {
      let suffix = ''
      if (!item.sameItems) {
        const name = item.to.name ? `[name="${item.to.name}"]` : ''
        const version =
          item.to.version ? `[version="${item.to.version}"]` : ''
        suffix = `${name}${version}`
      }
      if (item.to.importer && !item.from) {
        newQuery = `:project[name="${item.to.name}"]`
      } else if (item.from) {
        const fromName = `[name="${item.from.name}"]`
        const fromVersion =
          item.from.version ? `[version="${item.from.version}"]` : ''
        newQuery = `${fromName}${fromVersion} > :is(${query.trim()}${suffix})`
      } else {
        newQuery = `${query.trim()}${suffix}`
      }
    }
    updateQuery(newQuery)
    return undefined
  }

export const ResultItem = ({ item }: GridItemOptions) => {
  const updateQuery = useGraphStore(state => state.updateQuery)
  const query = useGraphStore(state => state.query)
  return (
    <div className="group relative z-10">
      {item.stacked ?
        <>
          {item.size > 2 ?
            <div className="transition-all group-hover:border-neutral-400 dark:group-hover:border-neutral-600 absolute border top-2 left-2 w-[97.5%] h-full bg-card rounded-lg"></div>
          : ''}
          <div className="transition-all group-hover:border-neutral-400 dark:group-hover:border-neutral-600 absolute border top-1 left-1 w-[99%] h-full bg-card rounded-lg"></div>
        </>
      : ''}

      {/* Card Top */}
      <Card
        renderAsLink={true}
        className={`relative group-hover:border-neutral-400 dark:group-hover:border-neutral-600 ${item.to ? 'cursor-pointer' : 'cursor-default'}`}
        onClick={onResultItemClick({ item, query, updateQuery })}>
        <CardHeader className="m-0 px-4 py-3 rounded-t-lg flex flex-row items-center justify-between border-b-[1px]">
          <TooltipProvider>
            <Tooltip>
              <TooltipTrigger className="truncate">
                <CardTitle className="text-md truncate">
                  {item.title}
                </CardTitle>
              </TooltipTrigger>
              <TooltipContent>
                <p>{item.title}</p>
              </TooltipContent>
            </Tooltip>
          </TooltipProvider>
          {item.type ?
            <div className="flex items-center justify-center text-pretty rounded-md border-[1px] px-2.5 py-0.5 text-xs text-muted-foreground">
              <p className="font-semibold">
                {item.stacked ? '' : <span>{item.type}</span>}
                <span className="font-medium px-1 text-muted-foreground">
                  dep of:
                </span>
                {item.stacked ?
                  <span>{item.size} packages</span>
                : <span>{stringifyNode(item.from)}</span>}
              </p>
            </div>
          : ''}
        </CardHeader>

        {/* Card Bottom */}
        <div className="flex flex-row px-3 py-2 items-center justify-between">
          {item.version ?
            <TooltipProvider>
              <div className="flex items-center justify-center rounded-sm border-[1px] px-2 py-1">
                <Tooltip>
                  <TooltipTrigger>
                    <p className="text-[0.65rem] text-muted-foreground font-mono font-medium truncate w-full m-0 p-0 align-baseline">
                      {item.version}
                    </p>
                  </TooltipTrigger>
                  <TooltipContent>
                    <p>{item.version}</p>
                  </TooltipContent>
                </Tooltip>
              </div>
            </TooltipProvider>
          : ''}
          <div className="flex gap-2 overflow-x-scroll">
            {item.labels?.length ?
              item.labels.map(i => (
                <div key={i}>
                  {
                    <Badge
                      className={labelClassNamesMap.get(i) || ''}>
                      {i}
                    </Badge>
                  }
                </div>
              ))
            : ''}
          </div>
        </div>
      </Card>
    </div>
  )
}
