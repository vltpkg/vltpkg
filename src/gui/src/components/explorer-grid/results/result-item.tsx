import { stringifyNode } from '@vltpkg/graph/browser'
import { useGraphStore } from '@/state/index.ts'
import { Badge } from '@/components/ui/badge.tsx'
import { Card, CardHeader } from '@/components/ui/card.tsx'
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from '@/components/ui/tooltip.tsx'
import { labelClassNamesMap } from '@/components/explorer-grid/label-helper.ts'
import { DataBadge } from '@/components/ui/data-badge.tsx'
import { Scale, EyeOff } from 'lucide-react'
import { ProgressCircle } from '@/components/ui/progress-circle.tsx'
import {
  getScoreColor,
  scoreColors,
} from '@/components/explorer-grid/selected-item/insight-score-helper.ts'
import { LICENSE_TYPES } from '@/lib/constants/index.ts'
import { cn } from '@/lib/utils.ts'

import type { MouseEvent } from 'react'
import type {
  GridItemData,
  GridItemOptions,
} from '@/components/explorer-grid/types.ts'
import type { ProgressCircleVariant } from '@/components/ui/progress-circle.tsx'
import type { Insights } from '@vltpkg/query'

export type ResultItemClickOptions = {
  item: GridItemData
  query: string
  updateQuery: (query: string) => void
}

const PackageOverallScore = ({
  className,
  insights,
}: {
  className?: string
  insights: Insights | undefined
}) => {
  if (!insights || !insights.scanned || !insights.score) return null

  const packageScore = insights.score
  const averageScore = packageScore.overall * 100
  const chartColor = getScoreColor(averageScore)
  const textColor = scoreColors[chartColor]

  return (
    <div className={className}>
      <ProgressCircle
        value={averageScore}
        variant={chartColor as ProgressCircleVariant}
        strokeWidth={5}
        className="size-7">
        <p
          className="font-mono text-xs font-medium tabular-nums"
          style={{ color: textColor }}>
          {averageScore}
        </p>
      </ProgressCircle>
    </div>
  )
}

const onResultItemClick =
  ({ item, query, updateQuery }: ResultItemClickOptions) =>
  (e: MouseEvent) => {
    e.preventDefault()
    if (!item.to) return
    let newQuery = ''
    if (item.stacked) {
      const name = item.to.name ? `[name="${item.to.name}"]` : ''
      const version = item.to.version ? `:v(${item.to.version})` : ''
      newQuery = `${query.trim()}${name}${version}`
    } else {
      let suffix = ''
      if (!item.sameItems) {
        const name = item.to.name ? `[name="${item.to.name}"]` : ''
        const version =
          item.to.version ? `:v(${item.to.version})` : ''
        suffix = `${name}${version}`
      }
      if (item.to.importer && !item.from) {
        newQuery = `:project[name="${item.to.name}"]`
      } else if (item.from) {
        const fromName = `[name="${item.from.name}"]`
        const fromVersion =
          item.from.version ? `:v(${item.from.version})` : ''
        newQuery = `${fromName}${fromVersion} > :is(${query.trim()}${suffix})`
      } else {
        newQuery = `${query.trim()}${suffix}`
      }
    }
    updateQuery(newQuery)
    return undefined
  }

export const ResultItem = ({ item }: GridItemOptions) => {
  const updateQuery = useGraphStore(state => state.updateQuery)
  const query = useGraphStore(state => state.query)
  const manifest = item.to?.manifest
  const insights = item.to?.insights
  return (
    <div className="group relative z-10">
      {item.stacked && (
        <>
          {item.size > 2 && (
            <div className="absolute left-2 top-2 h-full w-[97.5%] rounded-lg border bg-card transition-all group-hover:border-neutral-400 dark:group-hover:border-neutral-600"></div>
          )}
          <div className="absolute left-1 top-1 h-full w-[99%] rounded-lg border bg-card transition-all group-hover:border-neutral-400 dark:group-hover:border-neutral-600"></div>
        </>
      )}

      <Card
        renderAsLink
        className="duration-250 relative cursor-default transition-all group-hover:border-neutral-400 dark:group-hover:border-neutral-600"
        onClick={onResultItemClick({ item, query, updateQuery })}>
        <CardHeader className="relative flex w-full max-w-full flex-wrap items-baseline justify-between gap-3 px-3 py-2 md:flex-row">
          <div className="flex flex-col flex-wrap items-baseline gap-3 md:flex-row">
            {item.version && (
              <DataBadge
                variant="mono"
                classNames={{
                  wrapperClassName: 'md:-ml-1',
                  contentClassName: 'pt-0.5',
                }}
                tooltip={{ content: item.version }}
                content={item.version}
              />
            )}
            <TooltipProvider>
              <Tooltip>
                <TooltipTrigger className="grow cursor-default items-baseline justify-between overflow-hidden truncate text-left text-sm font-medium">
                  {item.title}
                </TooltipTrigger>
                <TooltipContent>{item.title}</TooltipContent>
              </Tooltip>
            </TooltipProvider>
          </div>

          <div className="flex flex-row flex-wrap items-center gap-3">
            {item.type && (
              <DataBadge
                classNames={{
                  valueClassName: 'lowercase',
                }}
                value={item.stacked ? '' : item.type}
                content={`dep of: ${item.stacked ? item.size : stringifyNode(item.from)}`}
              />
            )}
            {manifest?.private && (
              <DataBadge icon={EyeOff} content="Private Package" />
            )}
            {manifest?.license &&
              LICENSE_TYPES.some(
                i =>
                  i.toLowerCase() ===
                  manifest.license?.trim().toLowerCase(),
              ) && (
                <DataBadge
                  icon={Scale}
                  value={manifest.license}
                  content="License"
                />
              )}
            {manifest?.type && (
              <DataBadge
                content={manifest.type === 'module' ? 'ESM' : 'CJS'}
                classNames={{
                  wrapperClassName: 'h-[30px]',
                }}
              />
            )}
            <PackageOverallScore
              className="hidden md:flex"
              insights={insights}
            />
          </div>

          <div className="absolute -bottom-3.5 right-2.5 flex gap-2">
            {item.labels?.map(i => (
              <div key={i}>
                <Badge
                  className={cn(
                    labelClassNamesMap.get(i),
                    'h-[18px] rounded-full text-xxs font-normal text-inherit backdrop-blur-md',
                  )}>
                  {i}
                </Badge>
              </div>
            ))}
          </div>

          <PackageOverallScore
            className="absolute right-2 top-2 flex md:hidden"
            insights={insights}
          />
        </CardHeader>
      </Card>
    </div>
  )
}
