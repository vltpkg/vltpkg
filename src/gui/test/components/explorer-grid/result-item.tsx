import { describe, test, expect, afterEach, assert } from 'vitest'
import {
  cleanup,
  render,
  fireEvent,
  screen,
} from '@testing-library/react'
import html from 'diffable-html'
import { joinDepIDTuple } from '@vltpkg/dep-id'
import { type NodeLike } from '@vltpkg/graph'
import { useGraphStore as useStore } from '@/state/index.js'
import { ResultItem } from '@/components/explorer-grid/result-item.jsx'
import { type GridItemData } from '@/components/explorer-grid/types.js'

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

afterEach(() => {
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

test('ResultItem render with item', async () => {
  const item = {
    id: '1',
    labels: ['prod'],
    name: 'item',
    title: 'item',
    version: '1.0.0',
    sameItems: false,
    stacked: false,
    size: 1,
  } satisfies GridItemData
  render(<ResultItem item={item} />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

test('ResultItem render with type', async () => {
  const item = {
    id: '1',
    labels: ['prod'],
    type: 'prod',
    to: {} as NodeLike,
    from: {
      name: 'from',
      version: '1.0.0',
      id: joinDepIDTuple(['registry', '', 'from@1.0.0']),
    } as NodeLike,
    name: 'item',
    title: 'item',
    version: '1.0.0',
    sameItems: false,
    stacked: false,
    size: 1,
  } satisfies GridItemData
  render(<ResultItem item={item} />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

test('ResultItem render stacked two', async () => {
  const item = {
    id: '1',
    labels: ['dev'],
    name: 'two-edges-stacked',
    title: 'two-edges-stacked',
    version: '1.0.0',
    sameItems: false,
    stacked: true,
    size: 2,
  } satisfies GridItemData
  render(<ResultItem item={item} />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

test('ResultItem render stacked many', async () => {
  const item = {
    id: '1',
    labels: ['peer'],
    name: 'many-edges-stacked',
    title: 'many-edges-stacked',
    version: '1.1.1',
    sameItems: false,
    stacked: true,
    size: 5,
  } satisfies GridItemData
  render(<ResultItem item={item} />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

test('ResultItem render missing version and labels', async () => {
  const item = {
    id: '1',
    name: 'item',
    title: 'item',
    sameItems: false,
    stacked: false,
    size: 1,
  } as GridItemData
  render(<ResultItem item={item} />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

describe('ResultItem updates query store value', () => {
  test('stacked item', async () => {
    const item = {
      id: '1',
      labels: ['prod'],
      name: 'item',
      title: 'item',
      version: '1.0.0',
      sameItems: false,
      stacked: true,
      size: 2,
      to: {
        name: 'item',
        version: '1.0.0',
      } as NodeLike,
    } satisfies GridItemData
    render(<ResultItem item={item} />)

    fireEvent.click(screen.getByRole('link'))

    let query = ''
    const RetrieveQuery = () => (
      (query = useStore(state => state.query)), ''
    )
    render(<RetrieveQuery />)

    assert.deepEqual(
      query,
      ':root[name="item"][version="1.0.0"]',
      'should append item and version in order to narrow & show stacked items',
    )
  })

  test('importer item', async () => {
    const item = {
      id: '1',
      labels: ['prod'],
      name: 'my-item',
      title: 'my-item',
      version: '1.0.0',
      sameItems: false,
      stacked: false,
      size: 1,
      to: {
        name: 'my-item',
        version: '1.0.0',
        importer: true,
      } as NodeLike,
    } satisfies GridItemData
    render(<ResultItem item={item} />)

    fireEvent.click(screen.getByRole('link'))

    let query = ''
    const RetrieveQuery = () => (
      (query = useStore(state => state.query)), ''
    )
    render(<RetrieveQuery />)

    assert.deepEqual(
      query,
      ':project[name="my-item"]',
      'should prefix with :project followed by the name of the importer',
    )
  })

  test('has parent', async () => {
    const item = {
      id: '1',
      labels: ['prod'],
      name: 'item',
      title: 'item',
      version: '1.0.0',
      sameItems: false,
      stacked: false,
      size: 1,
      to: {} as NodeLike,
      from: {
        name: 'parent',
        version: '1.0.0',
      } as NodeLike,
    } satisfies GridItemData
    render(<ResultItem item={item} />)

    fireEvent.click(screen.getByRole('link'))

    let query = ''
    const RetrieveQuery = () => (
      (query = useStore(state => state.query)), ''
    )
    render(<RetrieveQuery />)

    assert.deepEqual(
      query,
      '[name="parent"][version="1.0.0"] > :is(:root)',
      'should prefix the parent name and version',
    )
  })

  test('add suffix', async () => {
    const item = {
      id: '1',
      labels: ['prod'],
      name: 'item',
      title: 'item',
      version: '1.0.0',
      sameItems: false,
      stacked: false,
      size: 1,
      to: {
        name: 'item',
        version: '1.0.0',
      } as NodeLike,
    } satisfies GridItemData
    render(<ResultItem item={item} />)

    fireEvent.click(screen.getByRole('link'))

    let query = ''
    const RetrieveQuery = () => (
      (query = useStore(state => state.query)), ''
    )
    render(<RetrieveQuery />)

    assert.deepEqual(
      query,
      ':root[name="item"][version="1.0.0"]',
      'should add name and version suffix',
    )
  })
})
