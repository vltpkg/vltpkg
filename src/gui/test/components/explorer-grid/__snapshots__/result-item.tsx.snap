// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResultItem render missing version and labels 1`] = `

<div>
  <div class="group relative z-10">
    <a
      href="#"
      class="block rounded-lg border bg-card text-card-foreground shadow-sm relative group-hover:border-neutral-400 dark:group-hover:border-neutral-600 cursor-default"
    >
      <div class="p-6 m-0 flex flex-row items-center justify-between rounded-t-lg border-b-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="truncate"
        >
          <h3 class="font-semibold tracking-tight text-md truncate">
            item
          </h3>
        </button>
      </div>
      <div class="flex h-12 w-full items-center justify-between gap-4 border-t-[1px] px-4 py-3">
        <div class="flex gap-2">
        </div>
      </div>
    </a>
  </div>
</div>

`;

exports[`ResultItem render stacked many 1`] = `

<div>
  <div class="group relative z-10">
    <div class="absolute left-2 top-2 h-full w-[97.5%] rounded-lg border bg-card transition-all group-hover:border-neutral-400 dark:group-hover:border-neutral-600">
    </div>
    <div class="absolute left-1 top-1 h-full w-[99%] rounded-lg border bg-card transition-all group-hover:border-neutral-400 dark:group-hover:border-neutral-600">
    </div>
    <a
      href="#"
      class="block rounded-lg border bg-card text-card-foreground shadow-sm relative group-hover:border-neutral-400 dark:group-hover:border-neutral-600 cursor-default"
    >
      <div class="p-6 m-0 flex flex-row items-center justify-between rounded-t-lg border-b-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="truncate"
        >
          <h3 class="font-semibold tracking-tight text-md truncate">
            many-edges-stacked
          </h3>
        </button>
      </div>
      <div class="flex h-12 w-full items-center justify-between gap-4 border-t-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="m-0 truncate rounded-sm border-[1px] px-2 py-1 align-baseline font-mono text-[0.65rem] text-muted-foreground"
        >
          1.1.1
        </button>
        <div class="flex gap-2">
          <div>
            <div class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-lime-100 text-neutral-900 hover:bg-lime-100/80">
              peer
            </div>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>

`;

exports[`ResultItem render stacked two 1`] = `

<div>
  <div class="group relative z-10">
    <div class="absolute left-1 top-1 h-full w-[99%] rounded-lg border bg-card transition-all group-hover:border-neutral-400 dark:group-hover:border-neutral-600">
    </div>
    <a
      href="#"
      class="block rounded-lg border bg-card text-card-foreground shadow-sm relative group-hover:border-neutral-400 dark:group-hover:border-neutral-600 cursor-default"
    >
      <div class="p-6 m-0 flex flex-row items-center justify-between rounded-t-lg border-b-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="truncate"
        >
          <h3 class="font-semibold tracking-tight text-md truncate">
            two-edges-stacked
          </h3>
        </button>
      </div>
      <div class="flex h-12 w-full items-center justify-between gap-4 border-t-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="m-0 truncate rounded-sm border-[1px] px-2 py-1 align-baseline font-mono text-[0.65rem] text-muted-foreground"
        >
          1.0.0
        </button>
        <div class="flex gap-2">
          <div>
            <div class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-fuchsia-100 text-neutral-900 hover:bg-fuchsia-100/80">
              dev
            </div>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>

`;

exports[`ResultItem render with item 1`] = `

<div>
  <div class="group relative z-10">
    <a
      href="#"
      class="block rounded-lg border bg-card text-card-foreground shadow-sm relative group-hover:border-neutral-400 dark:group-hover:border-neutral-600 cursor-default"
    >
      <div class="p-6 m-0 flex flex-row items-center justify-between rounded-t-lg border-b-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="truncate"
        >
          <h3 class="font-semibold tracking-tight text-md truncate">
            item
          </h3>
        </button>
      </div>
      <div class="flex h-12 w-full items-center justify-between gap-4 border-t-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="m-0 truncate rounded-sm border-[1px] px-2 py-1 align-baseline font-mono text-[0.65rem] text-muted-foreground"
        >
          1.0.0
        </button>
        <div class="flex gap-2">
          <div>
            <div class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-cyan-100 text-neutral-900 hover:bg-cyan-100/80">
              prod
            </div>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>

`;

exports[`ResultItem render with type 1`] = `

<div>
  <div class="group relative z-10">
    <a
      href="#"
      class="block rounded-lg border bg-card text-card-foreground shadow-sm relative group-hover:border-neutral-400 dark:group-hover:border-neutral-600 cursor-pointer"
    >
      <div class="p-6 m-0 flex flex-row items-center justify-between rounded-t-lg border-b-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="truncate"
        >
          <h3 class="font-semibold tracking-tight text-md truncate">
            item
          </h3>
        </button>
        <div class="flex items-center justify-center text-pretty rounded-md border-[1px] px-2.5 py-0.5 text-xs text-muted-foreground">
          <p class="font-semibold">
            <span>
              prod
            </span>
            <span class="px-1 font-medium text-muted-foreground">
              dep of:
            </span>
            <span>
              npm:from@1.0.0
            </span>
          </p>
        </div>
      </div>
      <div class="flex h-12 w-full items-center justify-between gap-4 border-t-[1px] px-4 py-3">
        <button
          data-state="closed"
          class="m-0 truncate rounded-sm border-[1px] px-2 py-1 align-baseline font-mono text-[0.65rem] text-muted-foreground"
        >
          1.0.0
        </button>
        <div class="flex gap-2">
          <div>
            <div class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-cyan-100 text-neutral-900 hover:bg-cyan-100/80">
              prod
            </div>
          </div>
        </div>
      </div>
    </a>
  </div>
</div>

`;
